nx/lib/nx.ex

```diff
@@ -7046,6 +7046,152 @@ defmodule Nx do
+    [
+      Nx.tensor([
+        [-0.5774, 0.8165, 0.0],
+        [-0.5774, -0.4082, 0.7071],
+        [-0.5774, -0.4082, -0.7071],
+        [0.0, 0.0, 0.0]
+      ]),
+      Nx.tensor([
+        [-1.7321, -4.0415, -2.3094],
+        [0.0, -4.0825, 3.266],
+        [0.0, 0.0, -2.8284]
+      ])
+    ]
+    iex(2)> Nx.dot(q, r) |> Nx.map(&Float.round(&1, 0))
+    Nx.tensor([
+      [1.0, -1.0, 4.0],
+      [1.0, 4.0, -2.0],
+      [1.0, 4.0, 2.0],
+      [0.0, -0.0, 0.0]
+    ])
+
+    iex(1)> [q, r] = Nx.qr(Nx.tensor([[1, -1, 4], [1, 4, -2], [1, 4, 2], [1, -1, 0]]), mode: :complete) |> Tuple.to_list() |> Enum.map(fn t -> Nx.map(t, &Float.round(&1, 4)) end)
+    [
+      Nx.tensor([
+        [-0.5, 0.5, -0.5, -0.5],
+        [-0.5, -0.5, 0.5, -0.5],
+        [-0.5, -0.5, -0.5, 0.5],
+        [-0.5, 0.5, 0.5, 0.5]
+      ]),
+      Nx.tensor([
+        [-2.0, -3.0, -2.0],
+        [0.0, -5.0, 2.0],
+        [-0.0, -0.0, -4.0],
+        [0.0, 0.0, 0.0]
+      ])
+    ]
+    iex(2)> Nx.dot(q, r) |> Nx.map(&Float.round(&1, 0))
+    Nx.tensor([
+      [1.0, -1.0, 4.0],
+      [1.0, 4.0, -2.0],
+      [1.0, 4.0, 2.0],
+      [1.0, -1.0, 0.0]
+    ])
+
+    iex(1)> {_q, _r, reflectors} = Nx.qr(Nx.tensor([[1, -1, 4], [1, 4, -2], [1, 4, 2], [1, -1, 0]]), mode: :raw)
+    iex(2)> Enum.map(reflectors, fn t -> Nx.map(t, &Float.round(&1, 4)) end)
+    [
+      Nx.tensor([
+        [1.0, 0.0, 0.0, 0.0],
+        [0.0, 1.0, 0.0, 0.0],
+        [0.0, 0.0, -0.6, 0.8],
+        [0.0, 0.0, 0.8, 0.6]
+      ]),
+      Nx.tensor([
+        [1.0, 0.0, 0.0, 0.0],
+        [0.0, -0.6667, -0.6667, 0.3333],
+        [0.0, -0.6667, 0.7333, 0.1333],
+        [0.0, 0.3333, 0.1333, 0.9333]
+      ]),
+      Nx.tensor([
+        [-0.5, -0.5, -0.5, -0.5],
+        [-0.5, 0.8333, -0.1667, -0.1667],
+        [-0.5, -0.1667, 0.8333, -0.1667],
+        [-0.5, -0.1667, -0.1667, 0.8333]
+      ])
+    ]
+
+  ## Error cases
+
+    iex> Nx.qr(Nx.tensor([[1, 1, 1, 1], [-1, 4, 4, -1], [4, -2, 2, 0]])) |> Tuple.to_list() |> Enum.map(fn t -> Nx.map(t, &Float.round(&1, 4)) end)
+    ** (ArgumentError) tensor must have at least as many rows than columns, got shape: {3, 4}
+  """
+  @doc type: :linalg
+  def qr(tensor, opts \\ []) do
+    %T{type: type, shape: shape} = tensor = tensor!(tensor)
+
+    assert_keys!(opts, [:mode])
+
+    opts = Keyword.merge([mode: :reduced], opts)
+
+    mode = opts[:mode]
+    valid_modes = [:reduced, :complete, :r, :raw]
+
+    unless mode in valid_modes do
+      raise ArgumentError,
+            "invalid :mode received. Expected one of #{valid_modes}, received: #{mode}"
+    end
+
+    output_type = Nx.Type.to_floating(type)
+
+    {q_shape, {k, n}} = Nx.Shape.qr(shape, opts)
+
+    q = iota(q_shape, type: output_type)
+
+    mode = opts[:mode]
+
+    r =
+      if mode == :reduced do
+        tensor[[0..(k - 1), 0..(n - 1)]]
+      else
+        tensor
+      end
+
+    impl!(tensor).qr(q, r, opts)
```

Comment:
So really when dispatching to the backend implementation you want to give it information about the output, the actual input, and then the options. So I wouldn't call out to other functions to build `q` and `r`. You should just provide the type/shape/name information, the input tensor, and the options and let the backend do the work. You can accomplish the `iota` and `slicing` from within `binary_backend`.

Reply:
I see! How would I go about defining `q` and `r` then? Should I build the `Tensor` struct by hand with empty data?

Reply:
You can do something like:

```elixir
q = %{tensor | shape: q_shape, names: q_names, type: output_type}
r = %{tensor | shape: r_shape, names: r_names, type: output_type}
```

It might help to think of the public facing API as a contract for shapes, types, and names. So all you should worry about doing is enforcing the shape, type, and name rules, and then let the backend worry about implementation details!